Add support for packet timestamping on Linux.

jdamato-fsly opened the pull request:
This change allows the user to specify the option "packet-timestamping" to turn
on packet timestamping on supported platforms (this change implements this for
Linux).

The data is exported in nanoseconds per thread via "duration-stats" (if enabled).

jdamato-fsly commented:
Offering this PR in place of a previous, very similar PR I opened (https://github.com/h2o/h2o/pull/2158).

I accidentally deleted my local git clone of the h2o repository where I created that original PR, so I am abandoning my original PR in favor of this new PR. My apologies for letting that original PR linger; I didn't have access to enough systems with new enough kernels to test that out.

TCP time stamping was added in kernel 4.14 (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=98aaa913b4ed250324429f0a9e6d5f77a3b5276c) - at the time of the first PR I only have 4.9 kernels to test on.

kazuho commented on src/main.c:
```diff
@@ -1343,14 +1351,28 @@ static void socket_reuseport(int fd)
     set_cloexec(fd);
 
+   {
+        int flag = 1;
```
Should we move this into `#if defined(__linux__)` below?

jdamato-fsly replied:
sure, thanks!

kazuho commented on src/main.c:
```diff
@@ -1343,14 +1351,28 @@ static void socket_reuseport(int fd)
+            }
+#else /* !defined(__linux__) */
+            fprintf(stderr, "[warning] cannot set SO_TIMESTAMPNS because the platform does not support it\n");
+#endif /* defined(__linux__) */
```
I like this approach (of emitting an warning on platforms that do not have support), and that coinsides with the idea of considering the concept of timestamping as platform-agnostic (please see my comment above).

kazuho commented on src/main.c:
```diff
@@ -1531,18 +1553,19 @@ static int on_config_listen(h2o_configurator_command_t *cmd, h2o_configurator_co
     yoml_t **ssl_node = NULL, **owner_node = NULL, **permission_node = NULL, **quic_node = NULL, **cc_node = NULL,
            **initcwnd_node = NULL;
     int proxy_protocol = 0;
+    int packet_timestamping_enabled = 0;
```
```suggestion
    int proxy_protocol = 0, packet_timestamping = 0;
```
`proxy_protocol` is a boolean, that turns on / off the feature. We can follow that convention by calling the new variable `packet_timestamping`.

jdamato-fsly replied:
sounds good, thanks.

kazuho commented on src/main.c:
```diff
@@ -1555,6 +1578,21 @@ static int on_config_listen(h2o_configurator_command_t *cmd, h2o_configurator_co
         if (proxy_protocol_node != NULL &&
             (proxy_protocol = (int)h2o_configurator_get_one_of(cmd, *proxy_protocol_node, "OFF,ON")) == -1)
             return -1;
+
+        if (packet_timestamping_node != NULL) {
+            if ((*packet_timestamping_node)->type != YOML_TYPE_SCALAR) {
+                h2o_configurator_errprintf(cmd, node, "`packet-timestamping` must be a string");
+                return -1;
+            }
+            if (strcasecmp((*packet_timestamping_node)->data.scalar, "ON") == 0) {
+                packet_timestamping_enabled = 1;
+            } else if (strcasecmp((*packet_timestamping_node)->data.scalar, "OFF") == 0) {
+                packet_timestamping_enabled = 0;
+            } else {
+                h2o_configurator_errprintf(cmd, node, "value of `packet-timestamping` must be either of: ON,OFF");
+                return -1;
+            }
+        }
```
```suggestion
        if (packet_timestamping_node != NULL &&
            (packet_timestamping = (int)h2o_configurator_get_one_of(cmd, *packet_timestamping_node, "OFF,ON")) == -1)
            return -1;
```

Proposed code is correct, but we do not need to check the type (as the argument of `h2o_configurator_parse_mapping` specifies the type as `:s`), and we can use `h2o_configurator_get_one_of` exactly the same way as we do for `proxy_protocol`.

jdamato-fsly replied:
updated this, thanks!

kazuho commented on src/main.c:
```diff
@@ -2720,7 +2761,7 @@ static void *run_loop(void *_thread_index)
                                       h2o_memcached_receiver);
 
-    if (conf.thread_map.entries[thread_index] >= 0) {
+	if (conf.thread_map.entries[thread_index] >= 0) {
```
I believe this is an unintentional change?

jdamato-fsly replied:
yes it is unintentional - sorry about that.

kazuho commented on include/h2o/socket.h:
```diff
@@ -110,6 +110,9 @@ struct st_h2o_socket_t {
     h2o_buffer_t *input;
+    int timestamping;
+    int needs_timestamp;
+
```
How about moving these variables right below `_skip_tracing`, with the types being changed to `unsigned : 1`?

Also, I would appreciate if we could better clarify the difference between the two flags (by changing their names, adding doc-comments, etc.), as the difference (or the necessity of having two) would not immediately be clear to people reading the code.

jdamato-fsly replied:
sure, i added a couple comments -- if you'd like me to change the names, as well, i am happy to do so.

kazuho commented on lib/common/socket/evloop.c.h:
```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
     while (1) {
         ssize_t rret;
         h2o_iovec_t buf = h2o_buffer_try_reserve(input, 4096);
+        size_t buflen = (buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 +1);
+
```
How about moving this below, like:
```c
struct iovec iov = { .iov_base = buf.base, .iov_len = buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 + 1 };`
```
The rationale is that we might want to keep memory allocation procedure (h2o_buffer_try_reserve) and failure handling close together

jdamato-fsly replied:
yes, i agree the failure handling and the allocation should be closer together. i fixed that.

i left buflen as a separate variable because i personally dislike extremely long / complex expressions (especially with ternaries) because i think they are hard to read... but ultimately up to you 😄  happy to change it to match your desired style.

kazuho commented on lib/common/socket/evloop.c.h:
```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
         }
-        while ((rret = read(fd, buf.base, buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 + 1)) == -1 && errno == EINTR)
+
+        /* if timestamping is enabled, lets setup everything we need to make it happen */
```
I think this comment does not apply here, but rather applies to the `if` block below?

kazuho commented on lib/common/socket/evloop.c.h:
```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
-        while ((rret = read(fd, buf.base, buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 + 1)) == -1 && errno == EINTR)
+
+        /* if timestamping is enabled, lets setup everything we need to make it happen */
+        size_t controllen = CMSG_SPACE(sizeof(struct timespec));
+        char control[controllen];
```
How about doing `char control[CMSG_SPACE(sizeof(struct timespec))]` and using `sizeof control` throughout, rather than having `controllen` as a separate variable?

jdamato-fsly replied:
i can make that change if you like -- i personally prefer separating stuff out for readability, but if you'd prefer it as you've described, i can make the change.

kazuho commented on lib/common/socket/evloop.c.h:
```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
+       return (int64_t)a->tv_sec * NSEC_PER_SEC + a->tv_nsec;
+}
+
+#if defined(__linux__) && !H2O_USE_LIBUV
```
```suggestion
#if defined(__linux__)
```
This file is evloop.c.h, therefore the second condition is always met :-)

jdamato-fsly replied:
ah right, copy paste bug 😄

kazuho commented on lib/common/socket/evloop.c.h:
```diff
@@ -236,7 +314,7 @@ static void read_on_ready(struct st_h2o_evloop_socket_t *sock)
         goto Notify;
 
-    if ((err = on_read_core(sock->fd, sock->super.ssl == NULL ? &sock->super.input : &sock->super.ssl->input.encrypted)) != NULL)
+    if ((err = on_read_core(sock)) != NULL)
```
I wonder if we could avoid the change of starting to pass `sock` to `on_read_core`.

Original code had the nice property of `on_read_core` being a freestanding function that could be evaluated without taking the rest of `st_h2o_evloop_socket_t` into consideration.

Perhaps what we want to consider is passing cmsg buffer as an optional argument to `on_read_core`, and call `handle_timestamp` from here after `on_read_core` returns.

jdamato-fsly replied:
OK i made this change. It feels slightly odd in the code - I suppose an alternative might be adding an on_timestamp callback or something instead. Let me know what you think.

kazuho commented on lib/common/socket/evloop.c.h:
```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
+        if (rret > 0) {
+            if (sock->super.needs_timestamp) {
+                handle_timestamp(sock, &msg);
+                sock->super.needs_timestamp = 0;
```
Is the intent here to collect timestamp once per each connection? If so, could you clarify what is the benefit for discarding the variables that have been collected mid-connection, please?

kazuho commented on lib/common/socket/evloop.c.h:
```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
+                ts = (struct scm_timestamping *)CMSG_DATA(cmsg);
+                packet_ts = timespec_to_nsec(&ts->ts[0]);
+
+                h2o_sliding_counter_start(&loop->packet_latency_nanosec_counter, packet_ts);
```
I'm not sure if I follow the logic here. IIUC, SO_TIMESTAMPNS is used to collect the latency between the NIC (or the driver) and userspace. But I do not see us using a timestamp collected within the userspace.

jdamato-fsly replied:
err, sorry about that! i think i made an error when moving code between branches. I have fixed this to obtain the userland timestamp for generating the sliding window.

kazuho reviewed (commented):
Yes, we should have observability into latencies. Thank you very much for working on that.

I see that the CI is failing (looks like some of the changes make h2o crash), but anyways below are my tentative review comments. Please let me know what you think.

jdamato-fsly commented on src/main.c:
```diff
@@ -1225,6 +1226,13 @@ static struct listener_config_t *add_listener(int fd, struct sockaddr *addr, soc
     }
     memset(&listener->ssl, 0, sizeof(listener->ssl));
     memset(&listener->quic, 0, sizeof(listener->quic));
+
+#if defined(__linux__)
+    listener->timestamping = timestamping;
+#else /* !defined(__linux__) */
+    listener->timestamping = 0;
+#endif /* defined(__linux__) */
+
```
sure this makes sense, thanks for the suggestion!

jdamato-fsly commented:
> I see that the CI is failing (looks like some of the changes make h2o crash), but anyways below are my tentative review comments. 

I'm not sure yet what exactly I broke -- I am looking into that now 🧐 

> Please let me know what you think.

Thanks for the detailed review! I pulled in most of the changes you suggested. I resolved most, but left a few conversations open in GitHub because I was not sure if you wanted further changes. If they are good to go, feel free to resolve them or let me know that further changes are needed.

Still working on the last 3 comments / conversations opened on the PR.

deweerdt commented on lib/common/socket/evloop.c.h:
```diff
@@ -115,9 +115,58 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
+
+        switch (cmsg->cmsg_type) {
+            case SO_TIMESTAMPNS:
+                ts = (struct scm_timestamping *)CMSG_DATA(cmsg);
```
I believe that for SO_TIMESTAMPNS the anciliary data is a `struct timespec`. This happens to work because the address of `&ts->ts[0]` is the same as `&ts` if `ts` were a `struct timespect`

jdamato-fsly replied:
ah yea you are right, thanks for pointing this out. i misread the docs. fixed in my latest commit.

jdamato-fsly commented:
> > I see that the CI is failing (looks like some of the changes make h2o crash), but anyways below are my tentative review comments.
> 
> I'm not sure yet what exactly I broke -- I am looking into that now 🧐
> 

Modified h2o to output some more debug output for the failing tests locally. Looks I am getting: `Socket operation on non-socket` which seems quite odd. The only way I think that could happen seems to be via some sort of memory corruption of the `sock` struct. I am still trying to track down what I've done wrong here.

jdamato-fsly commented:
> > > I see that the CI is failing (looks like some of the changes make h2o crash), but anyways below are my tentative review comments.
> > 
> > 
> > I'm not sure yet what exactly I broke -- I am looking into that now 🧐
> 
> Modified h2o to output some more debug output for the failing tests locally. Looks I am getting: `Socket operation on non-socket` which seems quite odd. The only way I think that could happen seems to be via some sort of memory corruption of the `sock` struct. I am still trying to track down what I've done wrong here.

Modified `on_read_core` to go back to the original PR style --- calling read sometimes and recvmsg other times. This is needed because sometimes non-socket FDs can make it here and recvmsg fails on non-socket FDs.

I've added some comments to explain this in the source, as well.
